python/pyspark/sql/column.py

```diff
@@ -129,6 +129,103 @@ def _(self, other):
     return _
 
 
+def _unresolved_named_lambda_variable(*name_parts):
+    """
+    Create o.a.s.sql.expressions.UnresolvedNamedLambdaVariable and
+    convert it to o.s.sql.Column
+
+    :param name_parts: str
+    """
+    sc = SparkContext._active_spark_context
+    name_parts_seq = _to_seq(sc, name_parts)
+    expressions = sc._jvm.org.apache.spark.sql.catalyst.expressions
+    return Column(
+        sc._jvm.Column(
+            expressions.UnresolvedNamedLambdaVariable(name_parts_seq)
+        )
+    )
+
+
+def _get_lambda_parameters(f):
+    import inspect
+
+    signature = inspect.signature(f)
+    parameters = signature.parameters.values()
+
+    # We should exclude functions that use
+    # variable args and keyword argnames
+    # as well as keyword only args
+    supported_parmeter_types = {
+        inspect.Parameter.POSITIONAL_OR_KEYWORD,
+        inspect.Parameter.POSITIONAL_ONLY,
+    }
+
+    # Validate that
+    # function arity is between 1 and 3
+    if not (1 <= len(parameters) <= 3):
```

Comment:
We do this because we know that none of the existing higher order functions support more than 3 arguments, correct?

What happens if I pass in a lambda that takes 3 arguments to a higher order function like `exists()` that only supports 1?

Reply:
Pretty much. Scala counterpart implement three overridden variants of `createLambda`:

https://github.com/apache/spark/blob/d69ed9afdf2bd8d03aaf835292b92692ec8189e9/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L3389-L3408 

I believe we could support arbitrary number of positional arguments here, as long as we generate `UnresolvedNamedLambdaVariable` placeholders, but I thought it is better to keep things aligned with Scala.

Reply:
> What happens if I pass in a lambda that takes 3 arguments to a higher order function like exists() that only supports 1?

This is handled by the planner, which should throw an exception like this:

```
: org.apache.spark.sql.AnalysisException: The number of lambda function arguments '2' does not match the number of arguments expected by the higher order function '1'.;
```

`AnalysisException` are nicely captured and re-thrown by PySpark, so it seems sufficient.

Reply:
So just to be clear, is this harmless?

```python
df.select(exists('numbers', lambda x, y, z: x < 0))
```

Reply:
Harmless, in a sense that it there is no correctness error, and it will fail with predictable exception (taken form live session)

```
AnalysisException                         Traceback (most recent call last)
...
AnalysisException: The number of lambda function arguments '3' does not match the number of arguments expected by the higher order function '1'.;
```

I guess the question is if this should fail proactively as well

```
exists('numbers', lambda x, y, z: x < 0)                                                                                                                                                                   
# Column<b'exists(numbers, lambdafunction((x < 0), x, y, z))'>
```

Reply:
If we decide it shoudl we could add `expected_nargs: Iterable[int]` to `_invoke_higher_order_function` and then pass it down to `_create_lambda` to fail if it doesn't match the signature.

Reply:
@nchammas After some consideration I've added early checks for arity ‒ even if analyzer does the right thing, it might be useful to fail early, if possible.

Reply:
Actually, I think the `AnalysisException` seems good enough for now. As long as the error message is readable to the users, less codes should be fine. I would rather just remove those function inspection in PySpark side.

Reply:
If you don't have strong opinion about it @HyukjinKwon, I'd keep it as is. While `AnalysisException` is good enough for seasoned Spark users, I see that many beginners are intimidated by JVM stack trace that comes with it, and often don't even try to figure out what's going on.

Also Python side checks, can prevent some naive mistakes early in cases when expressions are defined upfront. This is good, because it doesn't wait for the middle of the job to fail (obviously I'll add type hints on my side, to address that 😁), and there is really not much add code or complexity here ‒ we have to check arguments and arity anyway.

Reply:
I agree with @zero323 that for Python users getting a clean error thrown early on in Python is a much better experience than getting a huge JVM stack trace.

Reply:
I am not sure if this is a good way to duplicately handle error cases. I mean, for simple cechking, I am good with that but if it needs some considerable codes like the current, I am not sure yet.

I fully understand the concern about JVM stracktrace being exposed to Python users; however, I don't think we should add such checking to every API in PySpark. It will end up with duplicating whole [CheckAnalysis.scala](https://github.com/apache/spark/blob/290a528bff7bcb449714c1c6f1885bd0f804358d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala) and [TypeCoercion.scala](https://github.com/apache/spark/blob/297f406425d410e5c450a9fbe24679b49f00a553/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala).

I admit that I don't have a good idea except just fixing [utils.py#L76-L92](https://github.com/apache/spark/blob/0a95eb08003a115f59495b30aacaaa832940e977/python/pyspark/sql/utils.py#L76-L92) somehow better.

Hope we can discuss separately in a separate PR.

Reply:
> I am good with that but if it needs some considerable codes like the current, I am not sure yet.

Well, this change alone (without adding domain object that led to some line splitting) takes 7 lines of actual code  (excluding tests and docstings - 95dbda506cb3492c2e972e69be25eda4bf75b336), and that's only because I  was pretty generous with formatting. One or two liner would do just fine, if it wasn't for linter rules. Otherwise we just piggyback on signature analysis code - not something that can be delegated to the analyzer.

> I am not sure if this is a good way to duplicately handle error cases.

I'd argue that this doesn't really qualify as one. These specific analyzer errors target SQL expressions. Scala users will never get there, as this part will be handled by the compiler.

In general I anticipate that this feature might be somewhat confusing before users get used to it ‒ so safeguarding against common mistakes is a good idea, especially when predicted maintenance overhead is negligible.

Reply:
I actually pretty don't like such way (e.g., `_LambdaSpec(func=f, expected_nargs={1, 2})`) which looks a bit confusing. Also, to be complete, we should also check the input and output after `f` execution in Python side then..

The point is that, I think there might be another alternative for instance introducing type hint into PySpark, or somehow improving error messages at  [utils.py#L76-L92](https://github.com/apache/spark/blob/0a95eb08003a115f59495b30aacaaa832940e977/python/pyspark/sql/utils.py#L76-L92) as pointed out above. That's why I am currently thinking it might be better to do it separately.

Reply:
Not to reiterate https://github.com/apache/spark/pull/27433#discussion_r374069388 ‒ captured JVM exception is simply not as good as guest language exception. It cannot be inspected, it cannot be debugged and it doesn't directly connect to user objects. It doesn't really matter if how nicely it is wrapped and processed (please remember that these are interpreted languages and it is not unusual for user to actually drop to interactive debugger in the middle of the session).

If  consensus is that it is not the way to go I'll of course drop the checks. But please forgive me, if I won't do it without putting up a fight. 

> Also, to be complete, we should also check the input and output after f execution in Python side then..

I am not sure if I get the point. Input is generated so we know it is valid, and output type is validated. That's the best we can do until placeholders are resolved, which is out of Python's scope.

Reply:
This conversation reminds of the discussion on #16537. Perhaps we should discuss a general approach to Python-specific validation on the dev list?

I understand the maintainer perspective of not wanting to maintain duplicate logic, and at the same time agree that catching problems before things hit the JVM is just a better experience for Python users.

Reply:
> This conversation reminds of the discussion on #16537. Perhaps we should discuss a general approach to Python-specific validation on the dev list?

Yeah, I had a déjà vu feeling. Thanks for digging this up @nchammas.

General discussion is a great idea. I am aware that I am biased here, but I see people struggling  with where, when and how guest languages fail, almost every week, and it is really hard to get rid of the feeling that what we have at the moment is just not enough.

Reply:
Let's just run separately something we need to discuss separately. The codes here look, otherwise, pretty close to go.